aten/src/ATen/native/TensorTransformations.cpp

```diff
@@ -13,80 +13,132 @@ namespace native {
+
+  TensorOptions tensor_options =
+    TensorOptions(c10::kLong).
+    device(c10::kCPU);
+
+  return at::empty(new_shape, tensor_options);
 }
 
-Tensor flip_cpu(const Tensor& self, IntArrayRef dims) {
-  auto in_tensor = self;
-  const int64_t total_dims = in_tensor.dim();
-  auto flip_dims_b = at::dim_list_to_bitset(dims, total_dims);
-  Tensor out_tensor = at::empty_like(in_tensor, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-
-  // create contiguous strides for input tensor
-  auto stride_contiguous_v = std::vector<int64_t>(total_dims);
-  for (int64_t i = total_dims - 1; i >= 0; i--) {
-    if (i == total_dims - 1) {
-      stride_contiguous_v[i] = 1;
-    } else {
-      stride_contiguous_v[i] = std::max<int64_t>(in_tensor.size(i + 1), 1) * stride_contiguous_v[i + 1];
+
+std::vector<Tensor> build_indices_loop(Tensor input, IntArrayRef flip_dims) {
+  std::vector<Tensor> indices;
+  int64_t element_size_bytes = input.element_size();
+  for(auto dim: flip_dims) {
+    auto dim_size = input.size(dim);
+    auto index = build_index(input.ndimension(), dim, dim_size);
+    auto stride = input.stride(dim);
+    auto input_index_ptr = index.data_ptr<int64_t>();
+
+    for(int64_t i = 0; i < dim_size; i++) {
+      input_index_ptr[i] = static_cast<int64_t>(dim_size - i - 1) * stride * element_size_bytes;
     }
+    indices.push_back(index);
+  }
+  return indices;
+}
+
+static TensorIterator make_index_iterator(const Tensor input, const std::vector<Tensor> indices) {
+  TensorIteratorConfig config;
+  config.set_check_mem_overlap(false)
+        .check_all_same_dtype(false)
+        .declare_static_dtype_and_device(input.scalar_type(), input.device())
+        .add_output(Tensor())
+        .add_input(input);
+  for (auto& index : indices) {
+    config.add_input(index);
   }
+  return config.build();
+}
+
+template<typename scalar_t>
+void flip_cpu_kernel(TensorIterator& iter) {
+  int ntensor = iter.ntensors();
+  // When launch the index parallel version, set a relative samll grain size less than the INTERNAL::GRAIN_SIZE
+  // to make the whole available thread numbers get more balanced work load and a better cache location.
+  // The grain size here is chosen by the op benchmark to overcome the thread launch overhead
+  const int index_parallel_grain_size = 3000;
+  auto loop = [&](char** data, const int64_t* strides, int64_t n) {
+      auto indexer = Indexer(ntensor - 2, &data[2], &strides[2]);
+      char* dst = data[0];
+      char* src = data[1];
 
-  if (in_tensor.is_quantized()) {
-    AT_DISPATCH_QINT_AND_SUB_BYTE_TYPES(in_tensor.scalar_type(),
+      for (int64_t i = 0; i < n; i++) {
+          int64_t offset = indexer.get(i);
+          *(scalar_t*)(dst + strides[0] * i) = *(scalar_t*)(src + strides[1] * i + offset);
+      }
+  };
+
+  iter.for_each(loop, index_parallel_grain_size);
+}
+
+Tensor flip_cpu(const Tensor& self, IntArrayRef dims) {
+  auto input = self;
+
+  std::vector<int64_t> flip_dims;
+  for(int64_t dim: dims) {
```

Comment:
you don't check here that `dim` is in range, and you don't check that there are no repeating dims (previously it was done by `dim_list_to_bitset`).

Reply:
Yes, good catch! I think we should just use the `dim_list_to_bitset` as before but to build the indices from there